base/iterators.jl

```diff
@@ -780,16 +782,17 @@ length(P::ProductIterator) = prod(size(P))
 
 IteratorEltype(::Type{ProductIterator{Tuple{}}}) = HasEltype()
 IteratorEltype(::Type{ProductIterator{Tuple{I}}}) where {I} = IteratorEltype(I)
+
 function IteratorEltype(::Type{ProductIterator{T}}) where {T<:Tuple}
-    I = tuple_type_head(T)
-    P = ProductIterator{tuple_type_tail(T)}
-    IteratorEltype(I) == EltypeUnknown() ? EltypeUnknown() : IteratorEltype(P)
+    E = ntuple(n -> IteratorEltype(fieldtype(T, n)), _counttuple(T))
+    any(I -> I == EltypeUnknown(), E) && return EltypeUnknown()
+    return E[end]
 end
 
-eltype(::Type{<:ProductIterator{I}}) where {I} = _prod_eltype(I)
+eltype(::Type{<:ProductIterator{I}}) where {I} = @isdefined(I) ? _prod_eltype(I) : Any
 _prod_eltype(::Type{Tuple{}}) = Tuple{}
 _prod_eltype(::Type{I}) where {I<:Tuple} =
-    Base.tuple_type_cons(eltype(tuple_type_head(I)),_prod_eltype(tuple_type_tail(I)))
+    Tuple{ntuple(n -> eltype(fieldtype(I, n)), _counttuple(I))...}
```

Comment:
I didn't realize splatting in a `Tuple{...}` like this would be inferrable; that's great.